res/controllers/Numark-Scratch-scripts.js

```diff
@@ -0,0 +1,423 @@
+
+NumarkScratch.Deck = function(number) {
+    components.Deck.call(this, number);
+
+    const channel = number;
+
+    this.pflButton = new components.Button({
+        midi: [0x90 + channel, 0x1B],
+        key: "pfl",
+        output: function(value) {
+            const note = (value === 0x00 ? 0x80 : 0x90) + channel;
+            midi.sendShortMsg(note, 0x1B, this.outValueScale(value));
+        }
+    });
+
+    this.volume = new components.Pot({
+        inKey: "volume"
+    });
+
+    this.treble = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter3"
+    });
+
+    this.mid = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter2"
+    });
+
+    this.bass = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter1"
+    });
+
+    this.filter = new components.Pot({
+        group: "[QuickEffectRack1_" + this.currentDeck + "]",
+        inKey: "super1"
+    });
+
+    this.gain = new components.Pot({
+        inKey: "pregain"
+    });
+
+    const encoderInvert = function(key) {
+        return NumarkScratch.invertLoopEncoderFunction ? (key === "unshift" ? "shift" : "unshift") : key;
+    };
+
+    this.loopPlus = new components.Encoder({
+        midi: [0x93 + channel, 0x35],
+        [encoderInvert("unshift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "loop_double";
+        },
+        [encoderInvert("shift")]: function() {
+            this.group = "[Library]";
+            this.inKey = "MoveDown";
+        }
+    });
+
+    this.loopMinus = new components.Encoder({
+        midi: [0x93 + channel, 0x34],
+        [encoderInvert("unshift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "loop_halve";
+        },
+        [encoderInvert("shift")]: function() {
+            this.group = "[Library]";
+            this.inKey = "MoveUp";
+        }
+    });
+
+    this.loopButton = new components.Button({
+        midi: [0x93 + channel, 0x3F],
+        [encoderInvert("unshift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "beatloop_activate";
+        },
+        [encoderInvert("shift")]: function() {
+            this.group = "[Channel" + channel + "]";
+            this.inKey = "LoadSelectedTrack";
+        }
+    });
+
+    this.padSection = new NumarkScratch.PadSection(number);
+
+    this.shiftButton = new components.Button({
+        midi: [0x9F, 0x32],
+        input: function(channel, control, value) {
+            if (this.isPress(channel, control, value)) {
+                NumarkScratch.shift();
+                this.send(this.on);
+            } else {
+                NumarkScratch.unshift();
+                this.send(this.off);
+            }
+        }
+    });
```

Comment:
There is no per-deck shiftbutton afaict. So it would be better if this was directly part of the controller (just copy pasted into the `init` method). Then you can replace the explicit `midi.sendShortMsg` with a `shiftButton.trigger()` (improved encapsulation). 

```suggestion
    this.shiftButton = new components.Button({
        midi: [0x9F, 0x32],
        input: function(channel, control, value) {
            if (this.isPress(channel, control, value)) {
                NumarkScratch.shift();
                this.send(this.on);
            } else {
                NumarkScratch.unshift();
                this.send(this.off);
            }
        }
        trigger: function() {
            this.send(this.off);
        }
    });
```

Reply:
done